Index provider for Fulltext Indexes

ragadeeshu opened the pull request:
Not part of this pr:
- Consistency check support
- New kernel api support
- Core api support (await indexes)
- Randomized integration tests
- Recovery tests
- Deferred updates
- DbStructureCollector support
- TransactionLogAnalyzer support
- DumpStore support
- DumpCountsStore support
- Batchinserter tests

- [x] Make old `CommandReader` implementations not need the `IndexProviderMap` dependency, and instead hardcode them to whatever descriptor implementation was there before. See `RecordStorageCommandReaderFactory`
- [ ] Feature toggle to disable as much of the new code as possible.
- [ ] Schema store format migration.

ragadeeshu committed:
Reanme SchemaIndexProvider -> IndexProvider

ragadeeshu committed:
Rename IndexDescriptor -> SchemaIndexDescriptor

ragadeeshu committed:
Slightly split SchemaIndexDescriptor from IndexDescriptor

ragadeeshu committed:
Introduce FulltextIndexProvider

ragadeeshu committed:
Store and retrieve fulltext index in schema store

ragadeeshu committed:
Store and retrieve fulltext index, for real

ragadeeshu committed:
Properly store index names

ragadeeshu committed:
Query node from fulltext index

ragadeeshu committed:
No labels is all labels

ragadeeshu committed:
Introduce support for relationship indices

ragadeeshu committed:
Port fulltext tests, fix most bugs

ragadeeshu committed:
Store analyzer information as index metadata

ragadeeshu committed:
Fix most BloomIT tests

ragadeeshu committed:
Cleanup old fulltext code

ragadeeshu committed:
Fix more index update bugs

ragadeeshu committed:
Properly store index metadata, fix initial state

ragadeeshu committed:
Make everything compile

ragadeeshu committed:
Fix code style

ragadeeshu committed:
Fix a bunch of failing tests

ragadeeshu committed:
Fix a bunch of kernel tests

ragadeeshu committed:
All kernel tests pass

ragadeeshu committed:
Fix a bunch of TODOs

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

chrisvest committed:
Make the FulltextIndexProviderTest pass again after the merge.

chrisvest committed:
Cleanup in FulltextIndexProviderTest.

chrisvest committed:
Remove a system out print.

chrisvest committed:
Make BloomBackupIT pass.

chrisvest committed:
Fix some Causal Clustering tests after fulltext merge.

chrisvest committed:
Fix checkstyle errors.

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

ragadeeshu committed:
Fix bloomIT after merge

ragadeeshu committed:
Fix BuiltInProceduresAccepanceTest

ragadeeshu committed:
Fix consistency checker after fulltext integration

chrisvest committed:
Introduce an UNLABELLED label lock.

This lock is then used to coordinate schema modifications to e.g. fulltext indexes that cover the _arbitrary_ set of labels.
Without this lock, it is possible to concurrently drop an all-labels fulltext index while other transactions are trying to write to this index.
We _may_ be able to get rid of this lock again, if we serialise all fulltext updates to a dedicated writer thread, but that goes against the
grain of the current schema index architecture.

chrisvest committed:
Remove print-lines.

ragadeeshu committed:
Fix dumpcountsstoretest after fulltext merge

ragadeeshu committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

chrisvest committed:
More thorough testing of concurrent fulltext index access.

chrisvest committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

chrisvest committed:
Fix GraphDatabaseShutdownTest.

ragadeeshu committed:
Added some javadoc

ragadeeshu committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

ragadeeshu committed:
Remove a bunch of casts

ragadeeshu committed:
Fix another bunch of fulltext index todos

ragadeeshu committed:
Fix some more todos, clean up fulltext index code

ragadeeshu committed:
Remove FulltextIndexType

ragadeeshu committed:
More small polish for the fulltext addon

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

chrisvest committed:
Add TODOs after merge.

ragadeeshu committed:
Recovery for relationship indexes

ragadeeshu committed:
Fix more todos

chrisvest committed:
Make the resumable store copy refer to indexes via their index ids, instead of via their IndexDescriptors.

This is far more portable and robust, as custom IndexDescriptors allow increasingly exotic structure.

ragadeeshu committed:
Relaitonship index and all label index support for batch inserter

ragadeeshu committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

ragadeeshu committed:
Various small fixes and cleanup

ragadeeshu committed:
Remove an empty line

ragadeeshu committed:
Fix a bunch of failing tests

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

MishaDemianenko commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/tx/TxPullResponseDecoder.java:
```diff
@@ -42,8 +43,9 @@ protected void decode( ChannelHandlerContext ctx, ByteBuf msg, List<Object> out
     {
         NetworkReadableClosableChannelNetty4 logChannel = new NetworkReadableClosableChannelNetty4( msg );
         StoreId storeId = StoreIdMarshal.INSTANCE.unmarshal( logChannel );
+        //TODO empty index provider map, investigate if we need a rel one.
```
TODO

MishaDemianenko commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/token/ReplicatedTokenRequestSerializer.java:
```diff
@@ -120,8 +121,9 @@ public static ReplicatedTokenRequest unmarshal( ByteBuf buffer )
         ByteBuf txBuffer = Unpooled.wrappedBuffer( commandBytes );
         NetworkReadableClosableChannelNetty4 channel = new NetworkReadableClosableChannelNetty4( txBuffer );
 
+        //TODO empty index provider map, investigate if we need a rel one.
```
todo

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageCommandReaderFactory.java:
```diff
@@ -34,15 +35,15 @@
     private final CommandReader[] readers;
 
-    public RecordStorageCommandReaderFactory()
+    public RecordStorageCommandReaderFactory( IndexProviderMap indexProviderMap )
```
we should not update old command readers, also our command readers should not depend on whatever providers or indexes are available in a classpath since those are dynamically registered. We should be able to read our commands always by kernel only

MishaDemianenko commented on packaging/standalone/standalone-community/src/main/distribution/text/community/conf/neo4j.conf:
```diff
@@ -245,7 +245,7 @@ dbms.tx_log.rotation.retention_policy=1 days
 # org.neo4j.examples.server.unmanaged.HelloWorldResource.java from
 # neo4j-server-examples under /examples/unmanaged, resulting in a final URL of
-# http://localhost:#{default.http.port}/examples/unmanaged/helloworld/{nodeId}
+# http://localhost:#{default.http.port}/examples/unmanaged/helloworld/{entityId}
```
do not think this should be updated

ragadeeshu replied:
The dangers of intellij renaming.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/ConsistencyCheckTasks.java:
```diff
@@ -90,13 +91,13 @@
-    public List<ConsistencyCheckerTask> createTasksForFullCheck( boolean checkLabelScanStore, boolean checkIndexes,
-            boolean checkGraph )
+    public List<ConsistencyCheckerTask> createTasksForFullCheck( boolean checkLabelScanStore, boolean checkIndexes, boolean checkGraph,
+            IndexProviderMap indexes )
```
we should not pass IndexProviderMap here, we already have access to IndexAccessors here and should use those instead

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/IndexCheck.java:
```diff
@@ -38,7 +38,8 @@ public IndexCheck( IndexRule indexRule )
     public void check( IndexEntry record, CheckerEngine<IndexEntry, ConsistencyReport.IndexConsistencyReport> engine, RecordAccess records )
     {
-        int labelId = indexRule.schema().getLabelId();
+        //TODO support fulltext indexes in index consistency check
```
TODO and access by [0]

ragadeeshu replied:
This is the bleeding edge towards the not-yet-supported consistency checker.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -95,8 +96,8 @@ private void matchIndexesToNode(
-            long labelId = indexRule.schema().getLabelId();
-            if ( labels.contains( labelId ) )
+            int[] labelIds = indexRule.schema().getEntityTokenIds();
+            if ( Arrays.stream( labelIds ).asLongStream().anyMatch( labels::contains ) )
```
Instead of converting primitive int array to long objects to find matches please use primitive arrays instead (update NodeLabelReader to return primitives always as well)

klaren replied:
Also, the reason they differ is that `getEntityTokenIds` returns `int` and `NodeLabelReader` returns `long`. We should resolve this conflict where two different methods claim we have a different length of label ids.

ragadeeshu replied:
Arrays.stream with a int array creates a primitive int stream, so that should be fine, no? I updated NodeLabelReader to work with primitive sets.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java:
```diff
@@ -25,42 +25,48 @@
     {
         Dependencies deps = new Dependencies();
-        deps.satisfyDependencies( fileSystem, config, logService, pageCache, recoveryCollector, monitors );
+        //TODO figure out a way to not depend on procedures here, it is caused by the fulltext index.
```
TODO

ragadeeshu replied:
This will be resolved when we do the other consistency checker-work.

MishaDemianenko commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/SchemaDescriptorTranslation.scala:
```diff
@@ -21,15 +21,16 @@ package org.neo4j.cypher.internal.spi.v2_3
   implicit def kernelToCypher(index: KernelIndexDescriptor): SchemaTypes.IndexDescriptor =
     if (index.schema().getPropertyIds.length == 1)
-      SchemaTypes.IndexDescriptor(index.schema().getLabelId, index.schema().getPropertyId)
+    //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO

ragadeeshu replied:
See the other cypher comment.

MishaDemianenko commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/SchemaDescriptorTranslation.scala:
```diff
@@ -22,15 +22,16 @@ package org.neo4j.cypher.internal.spi.v3_1
   implicit def toCypher(index: KernelIndexDescriptor): SchemaTypes.IndexDescriptor = {
     assertSingleProperty(index.schema())
-    SchemaTypes.IndexDescriptor(index.schema().getLabelId, index.schema().getPropertyId())
+    //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO

ragadeeshu replied:
See the other cypher comment.

MishaDemianenko commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_3/TransactionBoundPlanContext.scala:
```diff
@@ -87,7 +88,8 @@ class TransactionBoundPlanContext(readOperationsSupplier: () => ReadOperations,
   private def getOnlineIndex(descriptor: KernelIndexDescriptor): Option[IndexDescriptor] =
     readOperationsSupplier().indexGetState(descriptor) match {
-      case InternalIndexState.ONLINE => Some(IndexDescriptor(descriptor.schema().getLabelId, descriptor.schema().getPropertyIds))
+      //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO

ragadeeshu replied:
See the other cypher comment.

MishaDemianenko commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundPlanContext.scala:
```diff
@@ -92,7 +93,8 @@ class TransactionBoundPlanContext(tc: TransactionalContextWrapper, logger: Inter
   private def getOnlineIndex(descriptor: KernelIndexDescriptor): Option[IndexDescriptor] =
     tc.statement.readOperations().indexGetState(descriptor) match {
-      case InternalIndexState.ONLINE => Some(IndexDescriptor(descriptor.schema().getLabelId, descriptor.schema().getPropertyIds))
+      //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO

ragadeeshu replied:
See the other cypher comment.

MishaDemianenko commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/IndexDescriptorCompatibility.scala:
```diff
@@ -20,24 +20,25 @@
 
   def kernelToCypher(index: KernelIndexDescriptor): CypherIndexDescriptor =
-    CypherIndexDescriptor(index.schema().getLabelId, index.schema().getPropertyIds)
+  //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO

ragadeeshu replied:
This is the bleeding edge towards the not-yet-supported cypher. I'm not sure that cypher will ever need this, but if that is the case we should probably figure out a nicer transition.

MishaDemianenko commented on community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/DbStructureLogicalPlanningConfiguration.scala:
```diff
@@ -63,8 +63,9 @@ object DbStructureLogicalPlanningConfiguration {
-  private def indexSet(indices: util.Iterator[Pair[String, Array[String]]]): Set[(String, Seq[String])] =
-    indices.asScala.map { pair => pair.first() -> pair.other().to[Seq] }.toSet
+  private def indexSet(indices: util.Iterator[Pair[Array[String], Array[String]]]): Set[(String, Seq[String])] =
+    //TODO we use a zero index here as to not bleed multi-token descriptors into cypher. At least for now.
```
TODO and the same [0] access pattern that should not be used

ragadeeshu replied:
See the other cypher comment.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedures.java:
```diff
@@ -47,93 +53,205 @@
+
+    private void await( String indexName ) throws IndexPopulationFailedKernelException
+    {
+        //TODO This is here because the core api really doesn't play nicely with multi token schema yet.
```
TODO

ragadeeshu replied:
This is the bleeding edge towards the not-yet-supported core api.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomProcedures.java:
```diff
@@ -47,93 +53,205 @@
+        }
+        catch ( SchemaRuleNotFoundException e )
+        {
+            return Stream.empty();
```
why error is empty stream? how do we recognize the difference between error and no result? the same question for almost all procedures in this class

ragadeeshu replied:
The exception we catch here is specifically schemarulenotfound, which means that we haven't created the bloom index yet (or we have deleted it), and thus we do not have any indexed properties and an empty stream is the correct result.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextAccessor.java:
```diff
@@ -17,18 +17,23 @@
+
+    Stream<String> propertyKeyStrings( IndexDescriptor descriptor );
+
+    IndexDescriptor indexDescriptorFor( String name, EntityType type, String[] entityTokens, String... properties );
```
always passes new String[0] as value for entityTokens in production code

ragadeeshu replied:
Yes, because bloom is the only production code we have right now, and that is the way they wanted it last time I asked. However this will not be the case for the fulltext surface when that arrives.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexAccessor.java:
```diff
@@ -0,0 +1,226 @@
+    @Override
+    public IndexReader newReader()
+    {
+        return IndexReader.EMPTY;
```
why do we return empty reader?

ragadeeshu replied:
Because the only use this will see is form consistency checker, and we do not support that yet.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexAccessor.java:
```diff
@@ -0,0 +1,226 @@
+        BoundedIterable<Document> empty = new BoundedIterable<Document>()
+        {
+            @Override
+            public long maxCount()
```
todo

ragadeeshu replied:
Yeah, consistency checker isn't supported yet, and this is to keep track of where the rough edges are.

ragadeeshu replied:
This will be seen to when we add consistency checker support.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexPopulator.java:
```diff
@@ -0,0 +1,151 @@
+    private final FulltextIndexDescriptor descriptor;
+    private final WritableFulltext index;
+
+    FulltextIndexPopulator( long indexId, FulltextIndexDescriptor descriptor, IndexSamplingConfig samplingConfig, WritableFulltext index )
```
half of the params are not used

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexPopulator.java:
```diff
@@ -0,0 +1,151 @@
+import org.neo4j.kernel.impl.api.index.sampling.IndexSamplingConfig;
+import org.neo4j.storageengine.api.schema.IndexSample;
+
+public class FulltextIndexPopulator implements IndexPopulator
```
this class is almost exact copy of LuceneIndexPopulator, looks like LuceneIndexPopulator should be used as superclass instead

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProvider.java:
```diff
@@ -0,0 +1,230 @@
+        }
+        catch ( TokenNotFoundException e )
+        {
+            throw new RuntimeException( e );
```
hm, runtime exception?

ragadeeshu replied:
Nice catch, we need to investigate this more!

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProvider.java:
```diff
@@ -0,0 +1,230 @@
+    @Override
+    public StoreMigrationParticipant storeMigrationParticipant( FileSystemAbstraction fs, PageCache pageCache )
+    {
+        return new SchemaIndexMigrator( fs, this );
```
migrate schema indexes? noooo

ragadeeshu replied:
At a first glance, it looked like the SchemaIndexMigrator would do the job just fine. Tests for this will be written at some point, but if you are certain that it won't work (why? :) ) we will look into it now.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProvider.java:
```diff
@@ -0,0 +1,230 @@
+
+    private class BadSchemaException extends IllegalArgumentException
+    {
+        BadSchemaException( String message )
```
bad schema error? with the only usage with a message that its impossible to index property?

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/FulltextFactory.java:
```diff
@@ -76,21 +72,14 @@ private Analyzer getAnalyzer( String analyzerClassName )
-        storageBuilder.withFileSystem( fileSystem ).withIndexFolder( indexRootFolder );
-        PartitionedIndexStorage storage = storageBuilder.build();
-        return new LuceneFulltext( storage, partitionFactory, properties, analyzer, identifier, type );
+        PartitionedIndexStorage storage = indexStorageFactory.indexStorageOf( indexId, config.get( GraphDatabaseSettings.archive_failed_index ) );
```
`indexStorageFactory.indexStorageOf( indexId, config.get( GraphDatabaseSettings.archive_failed_index ) )` should be extracted into separte method

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/FulltextQueryHelper.java:
```diff
@@ -17,18 +17,27 @@
+public class FulltextQueryHelper
 {
-    StubLuceneFulltext()
+    public static String createQuery( Collection<String> query, boolean fuzzy, boolean matchAll )
```
this whole method should be just rewritten and tests should be provided

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/LuceneFulltext.java:
```diff
@@ -17,73 +17,51 @@
-class LuceneFulltext extends AbstractLuceneIndex
+import static java.util.Collections.singletonMap;
+
+public class LuceneFulltext extends AbstractLuceneIndex implements Closeable
```
LuceneFulltext copies LuceneSchemaIndex behavior by copy-pasting, it should extend it instead.

ragadeeshu replied:
While I see where you are coming from, there is also a lot of stuff in LuceneSchemaIndex that we do not want, such as verifying uniqueness and dealing with IndexReader. The things that are identical are isOnline, markAsOnline and markAsFailed. Perhaps we should extract those to the AbstractLuceneIndex instead.

ragadeeshu replied:
Extracted.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/LuceneFulltextDocumentStructure.java:
```diff
@@ -49,18 +50,30 @@ private static DocWithId reuseDocument( long id )
     }
 
-    static Document documentRepresentingProperties( long id, Map<String,Object> values )
+    public static Document documentRepresentingProperties( long id, Map<String,Value> values )
```
this method is not used

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/ReadOnlyFulltext.java:
```diff
@@ -17,27 +17,17 @@
+    ScoreEntityIterator query( String query );
+
+    @Override
+    void close();
```
this interface overrides AutoCloseable close to pretend it does not throw anything while two implementations do that in reality and try to catch and rethrow exceptions while they do not need to do that since in a places where its used already we are free to throw IOExceptions since lucene do that anyway

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/SimpleFulltextReader.java:
```diff
@@ -85,28 +66,9 @@ public void close()
-        }
         Query query;
         try
         {
```
why query error is the same as empty result?

ragadeeshu replied:
Good catch, we need to put something better here.

ragadeeshu replied:
Upon further consideration, I think we should just throw this guy, since in some cases the query might be user-provided. If we are to catch it , we should do so in the bloom surface... Maybe.

MishaDemianenko commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/WritableFulltext.java:
```diff
@@ -56,43 +54,23 @@ public void drop() throws IOException
-    Set<String> getProperties()
-    {
-        return luceneIndex.getProperties();
+        luceneIndex.markAsOnline();
```
copy paste leak: locking needs to be used here as well

MishaDemianenko commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomFulltextIndexBenchmarks.java:
```diff
@@ -35,17 +35,13 @@
-
-@Ignore( "These are rudimentary benchmarks, but implemented via the jUnit framework to make them easy to run " +
-         "from an IDE." )
+@Ignore( "These are rudimentary benchmarks, but implemented via the jUnit framework to make them easy to run " + "from an IDE." )
```
i'm sure i already mentioned this before: BloomFulltextIndexBenchmarks should not be Ignored and live in the product. It should be moved into benchmarks repo and run there.

ragadeeshu replied:
Oh sorry, this is an old class that has just been moved around. These tests were used as an tool when developing the old fulltext index, will clean up.

MishaDemianenko commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProviderTest.java:
```diff
@@ -0,0 +1,390 @@
+import static org.junit.Assert.assertTrue;
+import static org.neo4j.graphdb.Label.label;
+
+public class FulltextIndexProviderTest
```
this class is a fountain of copy paste!

ragadeeshu replied:
Sorry, this test was a development tool that never got a proper cleanup, I'll see to it.

MishaDemianenko commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/lucene/LuceneFulltextTestSupport.java:
```diff
@@ -0,0 +1,226 @@
+
+    protected void await( IndexDescriptor fulltextIndexDescriptor ) throws IndexNotFoundKernelException
+    {
+        //TODO change this once you can use normal index await with multi-token indexes
```
TODO and continue inside is useless

ragadeeshu replied:
TODO beacuse of core api support being in the future. The continue is in order to pass check-style.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/exceptions/schema/SchemaRuleException.java:
```diff
@@ -33,9 +33,10 @@
-    protected final SchemaDescriptor descriptor;
     protected final String messageTemplate;
     protected final SchemaRule.Kind kind;
+    protected SchemaDescriptor descriptor;
```
two new fields should be final

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexProvider.java:
```diff
@@ -90,32 +93,10 @@
  * write to the index.
  */
-public abstract class SchemaIndexProvider extends LifecycleAdapter implements Comparable<SchemaIndexProvider>
+public abstract class IndexProvider<DESCRIPTOR extends IndexDescriptor> extends LifecycleAdapter implements Comparable<IndexProvider<?>>
```
`Comparable<IndexProvider<?>>` ? maybe `Comparable<IndexProvider<DESCRIPTOR>>` instead?

ragadeeshu replied:
I'm not sure, I think you should be able to compare different types of index providers?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/schema/index/IndexDescriptor.java:
```diff
@@ -36,15 +36,20 @@
         GENERAL,
-        UNIQUE
+        UNIQUE,
+        NON_SCHEMA
```
this should be renamed together with all related things to FULL_TEXT

ragadeeshu replied:
This isn't called full_text because it is not necessarily fulltext specific. I mean, it is right now, but there is nothing stating that is has to be, it could be any index that isn't a schema index. There is also no mention of fulltext anywhere in the kernel code as a whole, all of the fulltext aware code lives in the fulltext module, to me it feels like a nice thing that it doesn't bleed into the kernel code.

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/schema/SchemaDescriptor.java:
```diff
@@ -34,8 +35,15 @@
+    enum PropertySchemaType
+    {
+        SCHEMA_ALL_TOKENS,
+        NON_SCHEMA_ANY_TOKEN
```
this should be renamed as well

ragadeeshu replied:
Not sure I agree, but it definitely should be documented, very good point.

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/schema/SchemaDescriptor.java:
```diff
@@ -34,8 +35,15 @@
 {
+    boolean isAffected( long[] entityIds );
+
+    enum PropertySchemaType
```
would be nice to get javadoc for this class since the purpose of this enum is not really clear

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/schema/MultiTokenSchemaDescriptor.java:
```diff
@@ -0,0 +1,167 @@
+    @Override
+    public int keyId()
+    {
+        //TODO MultiTokenSchema support for the new kernel api. This method is only used in the AllStoreHolder
```
TODO

ragadeeshu replied:
This is the bleeding edge towards the not-yet-supported new kernel api.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -67,6 +67,7 @@
  */
 public interface ReadOperations
 {
+    long UNLABELLED = Integer.MAX_VALUE + 1L;
```
what is it for? there are no tests that check this as well

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/SchemaWriteOperations.java:
```diff
@@ -65,4 +67,7 @@ RelExistenceConstraintDescriptor relationshipPropertyExistenceConstraintCreate(
 
     void constraintDrop( ConstraintDescriptor constraint ) throws DropConstraintFailureException;
+
+    IndexDescriptor nonSchemaIndexCreate( IndexDescriptor nonSchemaIndexDescriptor )
```
nice try but no, please rename the whole stack of those methods

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java:
```diff
@@ -892,7 +893,8 @@ public RelationshipResult( Relationship relationship )
         NODE_LABEL_PROPERTY( "node_label_property" ),
-        NODE_UNIQUE_PROPERTY( "node_unique_property" );
+        NODE_UNIQUE_PROPERTY( "node_unique_property" ),
+        NON_SCHEMA( "non_schema" );
```
this should be renamed to full text

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/EntityUpdates.java:
```diff
@@ -29,67 +29,69 @@
  * This is of use in populating indexes that might be relevant to node label and property combinations.
  */
-public class NodeUpdates implements PropertyLoader.PropertyLoadSink
+public class EntityUpdates implements PropertyLoader.PropertyLoadSink
```
class was heavily updated but there are no new tests

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java:
```diff
@@ -272,25 +270,35 @@ public void start() throws Exception
+                }
+            }
+
+            if ( !rebuildingNodeDescriptors.isEmpty() )
```
i'm sure that starting those 2 jobs as they are now will not gonna play well since each of those will gonna try to perform multi-threaded index population with tons of threads fighting with each other for no reason

ragadeeshu replied:
In what sense will they fight? They scan different stores and apply updates to different indexes?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMap.java:
```diff
@@ -45,10 +48,14 @@
 public final class IndexMap implements Cloneable
```
class was heavily updated but again no new tests

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexStoreView.java:
```diff
@@ -41,26 +42,39 @@
      * @return node updates container
      */
-    NodeUpdates nodeAsUpdates( long nodeId );
+    //TODO This seems to only be used in tests
```
TODO again

ragadeeshu replied:
Right, this todo here... It only seems to be used in tests, but as far as I can tell, it has been that way for quite a while, and since we did not cause it, only noticed it, we did not prioritize fixing it. Would you rather see that we took time to remove it, removed the todo, or neither?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -297,6 +312,7 @@ void flipAfterPopulation()
         {
             try
             {
+                Thread.sleep( 1000 );
```
yes right, we already too fast lets give a cpu some time to sleep :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -317,9 +333,14 @@ private IntStream propertyKeyIds( IndexPopulation population )
-        return populations.stream().mapToInt( population -> population.schema().getLabelId() ).toArray();
+        if ( populations.stream().anyMatch( indexPopulation -> indexPopulation.schema().getEntityTokenIds().length == 0 ) )
+        {
+            //No token is any token
```
this should not be here, empty array and [0] element of array is leaking everywhere

ragadeeshu replied:
Do you have any proposals for better defining any label/reltype than the empty array? I saw it being defined as empty array somewhere in existing code, and decided to follow that pattern since I couldn't figure out anything better.

ragadeeshu replied:
Replaced with a constant for now, feels a bit better at least.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java:
```diff
@@ -371,9 +372,11 @@ public long nodeGetFromUniqueIndexSeek(
         assertPredicatesMatchSchema( index.schema(), predicates );
-        int labelId = index.schema().getLabelId();
+        // This is fine because of the assertion above
+        int labelId = index.schema().getEntityTokenIds()[0];
```
same leak here

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java:
```diff
@@ -502,6 +529,64 @@ public String indexGetFailure( Statement state, IndexDescriptor descriptor )
         return schemaReadDelegate.indexGetFailure( state, descriptor );
     }
 
+    private void lockSchemaExclusive( KernelStatement state, SchemaDescriptor schema )
```
what this 2 new methods are doing? i can't see any tests for them as well?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/SchemaImpl.java:
```diff
@@ -132,7 +133,8 @@ private IndexDefinition descriptorToDefinition( final ReadOperations statement,
         try
         {
-            Label label = label( statement.labelGetName( index.schema().getLabelId() ) );
+            //TODO we use a zero index here to avoid bleeding into core api. We need to figure out how to deal with the core api in regards to fulltext indexes.
```
TODO again

ragadeeshu replied:
This is the bleeding edge towards the not-yet-supported core api.

MishaDemianenko reviewed (changes requested):
First bunch of comments:
* also FulltextIndexUpdater is almost a copy of LuceneIndexUpdater;
* GraphDatabaseFacade contains a hack
* package of whole thing is `org.neo4j.kernel.api.impl.fulltext.*`
which its not, have nothing to do with kernel
* lots of places do something like `descriptor.getIndexDescriptor().schema().entityType() == EntityType.NODE` and then operate on top of descriptor itself - so again leak of abstraction there but i'm not done with that part fully so will only mention it as general note
* there are no consistency checker updates for new indexes?
* now changes of analyzer property will gonna drop all of the full text indexes. That behaviour is different compared to other indexes and we need to warn about that in property description and write warning messages into the log file at least.
*we talked about flag to disable the whole thing 2 days ago and i do not see it. Where is it?

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexEntryUpdate.java:
```diff
@@ -35,7 +36,7 @@
  *
  * @param <INDEX_KEY> {@link LabelSchemaSupplier} specifying the schema
```
`SchemaDescriptorSupplier`

ragadeeshu replied:
fixed

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexProvider.java:
```diff
@@ -90,32 +93,10 @@
  * write to the index.
  */
-public abstract class SchemaIndexProvider extends LifecycleAdapter implements Comparable<SchemaIndexProvider>
+public abstract class IndexProvider<DESCRIPTOR extends IndexDescriptor> extends LifecycleAdapter implements Comparable<IndexProvider<?>>
```
I think this change introduces too many raw types. Could we look into removing the generic type?

klaren commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/schema/MultiTokenSchemaDescriptor.java:
```diff
@@ -17,19 +17,8 @@
-import org.neo4j.kernel.api.index.SchemaIndexProvider;
-
-public interface SchemaIndexProviderMap extends Function<SchemaIndexProvider.Descriptor, SchemaIndexProvider>
+public interface MultiTokenSchemaDescriptor extends SchemaDescriptor
```
A javadoc describing what multi token is would be nice.

ragadeeshu committed:
Reomove some redundant generics form IndexProvider

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java:
```diff
@@ -68,6 +69,7 @@
  */
 public interface ReadOperations
 {
+    long UNLABELLED = Integer.MAX_VALUE + 1L;
```
This looks dangerous. The `-1` is "protected" by `org.neo4j.kernel.impl.store.id.IdGeneratorImpl#INTEGER_MINUS_ONE` and `org.neo4j.kernel.impl.store.id.validation.IdValidator#isReservedId` (we already have a card to make this relationship clear). Might be better to have it `Long.MIN_VALUE` or something not in the path of future label ids if we migrate to `long` for that.

ragadeeshu committed:
Clean up some fulltext testing

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/RelationshipStoreScan.java:
```diff
@@ -0,0 +1,222 @@
+        {
+            return true;
+        }
+        for ( long candidate : relTypeFilter )
```
`int`

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/RelationshipStoreScan.java:
```diff
@@ -0,0 +1,222 @@
+import static java.util.Collections.emptyIterator;
+import static org.neo4j.kernel.impl.store.record.RecordLoad.FORCE;
+
+public class RelationshipStoreScan<FAILURE extends Exception> implements StoreScan<FAILURE>
```
Lots of duplicate code here, e.g `acceptUpdate`, `getProgress`, `valueOf` etc. is dientical to `StoreViewNodeStoreScan`.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexProviderMap.java:
```diff
@@ -0,0 +1,66 @@
+import org.neo4j.kernel.api.schema.index.IndexDescriptor;
+import org.neo4j.kernel.api.schema.index.SchemaIndexDescriptor;
+
+public interface IndexProviderMap extends Function<IndexProvider.Descriptor,IndexProvider>
```
Really no need to extend `Function` here.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexProviderMap.java:
```diff
@@ -0,0 +1,66 @@
+    };
+
+    @Override
+    IndexProvider apply( IndexProvider.Descriptor descriptor ) throws IndexProviderNotFoundException;
```
rename

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -93,21 +96,23 @@
     private final IndexStoreView storeView;
     private final LogProvider logProvider;
     protected final Log log;
+    private EntityType type;
```
`final`

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -326,9 +338,14 @@ private IntStream propertyKeyIds( IndexPopulation population )
+        if ( populations.stream().anyMatch( indexPopulation -> indexPopulation.schema().getEntityTokenIds().length == 0 ) )
+        {
+            //No token is any token
+            return new int[0];
```
At least make this a constant with a proper name and remove the comment.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/PropertyCommandsExtractor.java:
```diff
@@ -105,6 +112,17 @@ public boolean visitPropertyCommand( PropertyCommand command )
+        }
         return false;
     }
 
```
`containsAnyNodeOrPropertyUpdate` affects relationships new as well. Rename to reflect this.

ragadeeshu replied:
updated

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java:
```diff
@@ -371,9 +372,11 @@ public long nodeGetFromUniqueIndexSeek(
         assertIndexOnline( state, index );
         assertPredicatesMatchSchema( index.schema(), predicates );
-        int labelId = index.schema().getLabelId();
+        // This is fine because of the assertion above
```
Make this explicit. E.g change `assertIndexHasOneEntityToken` to `getOneLableIdOrThrow` and return the id from there.

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/SchemaStore.java:
```diff
@@ -42,7 +43,7 @@
 import static org.neo4j.kernel.impl.store.record.RecordLoad.CHECK;
 
-public class SchemaStore extends AbstractDynamicStore implements Iterable<SchemaRule>
+public class SchemaStore extends AbstractDynamicStore// implements Iterable<SchemaRule>
```
Implementing comments is not supported :wink:

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/DefaultIndexProviderMap.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import org.neo4j.internal.kernel.api.IndexCapability;
```
Unused

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/DefaultIndexProviderMap.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    @Override
+    public IndexProvider getDefaultSchemaIndexProvider()
```
Changes return type from specified to raw.

klaren commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -521,31 +519,59 @@ private void repopulateAllIndexes() throws IOException, IndexEntryConflictExcept
+        if ( nodeDescriptors.stream().anyMatch( descriptor -> descriptor.schema().getEntityTokenIds().length == 0 ) )
+        {
+            //No token is any token
+            return new int[0];
```
Constant

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:
```diff
@@ -726,19 +726,31 @@ public void constraintDrop( KernelStatement state, ConstraintDescriptor constrai
+    @Override
+    public IndexDescriptor indexGetForName( KernelStatement state, String name )
+    {
+        // This is only used for querying, and thus is is "fine" (at least for now) that it is not tx-state aware
```
At least assert that `state.hasTxStateWithChanges()` is false and throw otherwise.

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextConfig.java:
```diff
@@ -17,30 +17,21 @@
 import static org.neo4j.kernel.configuration.Settings.setting;
 
 /**
  * Configuration parameters for the bloom fulltext addon.
```
Still referencing bloom.

ragadeeshu committed:
Some smaller fulltext cleanup

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProvider.java:
```diff
@@ -0,0 +1,230 @@
+        accessorsByName = new HashMap<>();
+    }
+
+    public ScoreEntityIterator query( IndexDescriptor descriptor, String query ) throws IOException
```
Should probably be `FulltextIndexProvider` for the parameter as the map expects that.

ragadeeshu replied:
I did it the other way around, hope that works for you

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/integrations/kernel/FulltextIndexProviderFactory.java:
```diff
@@ -0,0 +1,90 @@
+@Service.Implementation( KernelExtensionFactory.class )
+public class FulltextIndexProviderFactory extends KernelExtensionFactory<FulltextIndexProviderFactory.Dependencies>
+{
+    public static final String KEY = "fulltext";
```
`private`

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/FulltextFactory.java:
```diff
@@ -17,48 +17,44 @@
-     * @param fileSystem The filesystem to use.
-     * @param storeDir Store directory of the database.
      * @param analyzerClassName The Lucene analyzer to use for the {@link LuceneFulltext} created by this factory.
+     * @param config
```
Missing javadoc

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/FulltextFactory.java:
```diff
@@ -17,48 +17,44 @@
-     * @param storeDir Store directory of the database.
      * @param analyzerClassName The Lucene analyzer to use for the {@link LuceneFulltext} created by this factory.
+     * @param config
      * @throws IOException
```
Does it throw?

klaren commented on enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/lucene/LuceneFulltextFieldEncoding.java:
```diff
@@ -17,7 +17,7 @@
  * You should have received a copy of the GNU Affero General Public License
```
This class can be inlined, the static method has 1 usage which only pass the parameters as is.

klaren commented on enterprise/fulltext-addon/src/main/resources/META-INF/services/org.neo4j.kernel.extension.KernelExtensionFactory:
```diff
@@ -1 +1 @@
-org.neo4j.kernel.api.impl.fulltext.integrations.bloom.BloomKernelExtensionFactory
\ No newline at end of file
+org.neo4j.kernel.api.impl.fulltext.integrations.kernel.FulltextIndexProviderFactory
```
missing newline

ragadeeshu replied:
fixed

klaren commented on enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/integrations/bloom/BloomBackupIT.java:
```diff
@@ -55,6 +56,8 @@
 {
```
Is this bloom specific or could it be renamed to fulltext?

ragadeeshu replied:
Bloom name makes sense, as it uses the bloom surface.

klaren reviewed (changes requested):
My first iteration.

General comments:
* We talked about having the full-text index be configured to be eventually consistent or not. Is this planned?
* "Non-schema" that is accessed through the schema makes me a bit confused. Maybe we can come up with a better name.
* Raw types should be avoided in new code. Some effort needs to be dedicated to this.
* Still a lot of TODOs left, I did not mention them since I'm assuming those will be acted upon before this is merged.
* Should we keep the bloom name and have it live side by side or migrate bloom to use the final surface later?

ragadeeshu committed:
More small fulltext-polish

ragadeeshu committed:
Cosmetic indexing changes.

ragadeeshu committed:
Merge common items for Node and Rel scanners

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

chrisvest committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

ragadeeshu committed:
Minor javadoc fix

chrisvest committed:
Merge branch '3.4' into 3.4-fulltext-integration

ragadeeshu committed:
Remove a bunch of duplicated lucene index code

ragadeeshu committed:
Lucene fulltext cleanup and renames

ragadeeshu committed:
Clean up fulltext index accessor

ragadeeshu committed:
FulltextIndexProvider cleanup

ragadeeshu committed:
Merge branch '3.4-fulltext-integration' of github.com:ragadeeshu/neo4j into 3.4-fulltext-integration

ragadeeshu committed:
Split out the bloom addon to its own module

ragadeeshu committed:
Minor test fixes

ragadeeshu committed:
Some exception improvements

ragadeeshu committed:
Add new tests for EntityUpdates

ragadeeshu committed:
Tests and polish for IndexMap

ragadeeshu committed:
Use proper constant in IndexMapTest

ragadeeshu committed:
Merge branch '3.4' into 3.4-fulltext-integration

ragadeeshu committed:
Fix an unfortunate recursion

ragadeeshu committed:
Merge branch '3.4' into 3.4-fulltext-integration

ragadeeshu committed:
First part of fulltext consistency checker support

ragadeeshu committed:
Initial fulltext cluster support

ragadeeshu committed:
Builder for IndexRule

ragadeeshu committed:
Some more cluster tests

ragadeeshu committed:
Fix mocking in IndexingServiceTest

ragadeeshu committed:
Fix a bunch of TODOs

fickludd committed:
Made SchemaIndexDescriptor a IndexReference and changed indexCreate

fickludd committed:
Made SchemaRule an interface

fickludd committed:
Made IndexRule extend SchemaIndexDescriptor

fickludd committed:
Remove unused argument

ragadeeshu committed:
First draft for fulltext surface, kinda wip

fickludd committed:
Rename

SchemaIndexDescriptor => PendingIndexDescriptor
IndexRule => IndexDescriptor
SchemaIndexDescriptorFactory => IndexDescriptorFactor

ragadeeshu committed:
Use real IndexDescriptors in indexing

ragadeeshu committed:
Make IndexMeta extends IndexDescriptor

ragadeeshu committed:
Move providermethods for indexreference

ragadeeshu committed:
Rename IndexMeta to CapableIndexReference

fickludd committed:
Remove redundant methods on IndexProxy

fickludd committed:
Let SchemaCache store CapableIndexDescriptor

fickludd committed:
Rename IndexDescriptor => StoreIndexDescriptor

fickludd committed:
Rename PendingIndexDescriptor => IndexDescriptor

fickludd committed:
Remove DefaultIndexReference static constructors

fickludd committed:
Remove DefaultIndexReference

ragadeeshu committed:
Remove old imports

ragadeeshu committed:
Cleanup around IndexDescriptor

ragadeeshu committed:
Remove DefaultCapableIndexReference

ragadeeshu committed:
Remove sortByType in IndexDescriptor

ragadeeshu committed:
Merge CapableIndexReference with IndexReference

ragadeeshu committed:
Move CapableIndexDescriptor

ragadeeshu committed:
Cleanup Descriptor classes

ragadeeshu committed:
Move forLabel methods to TestIndexDescriptorFactory

Also, fix the use of name()

fickludd committed:
Rework StoreIndexDescriptor constructors

fickludd committed:
Remove StoreIndexDescriptor.indexRule constructor

fickludd committed:
indexUniqueCreate method for ConstraintIndexCreator

fickludd committed:
Remove StoreIndexDescriptor.constraintRule constructor

fickludd committed:
Fix some test failures

ragadeeshu committed:
Fix a bunch of failing tess

ragadeeshu committed:
Update DBstructureCollector

ragadeeshu committed:
More test fixes

ragadeeshu committed:
Remove SchemaDescriptor from IndexProxy

ragadeeshu committed:
Fix ConstraintIndexCreatorTest

ragadeeshu committed:
Code style fixes

ragadeeshu committed:
Test and fix supplied index provider

ragadeeshu committed:
Minor cleanups

ragadeeshu committed:
Naming cleanups

ragadeeshu committed:
Checkstyle fixes

ragadeeshu committed:
Merge branch '3.5-index-management-cleanup' into 3.4-fulltext-integration

# Conflicts:
#	community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/ConsistencyCheckTasks.java
#	community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/CountsBuilderDecorator.java
#	community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java
#	community/consistency-check/src/main/java/org/neo4j/consistency/checking/index/IndexAccessors.java
#	community/consistency-check/src/test/java/org/neo4j/consistency/checking/SchemaRuleUtil.java
#	community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java
#	community/consistency-check/src/test/java/org/neo4j/consistency/report/ConsistencyReporterTest.java
#	community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/SchemaDescriptorTranslation.scala
#	community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_3/TransactionBoundPlanContext.scala
#	community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/SchemaDescriptorTranslation.scala
#	community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundPlanContext.scala
#	community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/IndexDescriptorCompatibility.scala
#	community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundPlanContext.scala
#	community/kernel/src/main/java/org/neo4j/kernel/api/ReadOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/SchemaWriteOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexPopulator.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/index/IndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/index/LoggingMonitor.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/schema/SchemaDescriptorFactory.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/schema/index/IndexDescriptor.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/schema/index/SchemaIndexDescriptorFactory.java
#	community/kernel/src/main/java/org/neo4j/kernel/api/txstate/TransactionState.java
#	community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/BuiltInProcedures.java
#	community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/DataIntegrityValidatingStatementOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/AbstractDelegatingIndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/AbstractSwallowingIndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/EntityUpdates.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/FlippableIndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMap.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMapReference.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexMeta.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexProviderMap.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexProxyCreator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexStoreView.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/NodeUpdatesIterator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/OnlineIndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/PopulatingIndexProxy.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/PropertyCommandsExtractor.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/PropertyLoader.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/PropertyPhysicalToLogicalConverter.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/RebuildingIndexDescriptor.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingController.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingControllerFactory.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/operations/CountsOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/operations/EntityReadOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/operations/SchemaReadOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/operations/SchemaWriteOperations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/NodeSchemaMatcher.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/ConstraintIndexCreator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/IndexTxStateUpdater.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/state/TxState.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DefaultCapableIndexReference.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/DefaultIndexReference.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/SchemaCache.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorageLayer.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexAccessor.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaIndexPopulator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NumberIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexAccessor.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexFiles.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexPopulator.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexProvider.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/AllStoreHolder.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/IndexTxStateUpdater.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/NodeSchemaMatcher.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Operations.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/RecordStorageEngine.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/impl/recordstorage/TransactionToRecordStateVisitor.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/store/SchemaStorage.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/IndexRule.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/SchemaRuleDeserializer2_0to3_1.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/store/record/SchemaRuleSerialization.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/IndexBatchTransactionApplier.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/IndexUpdatesWork.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/DefaultIndexProviderMap.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/DirectIndexUpdates.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/IndexUpdates.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/OnlineIndexUpdates.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/DynamicIndexStoreView.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/NeoStoreIndexStoreView.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/NodeStoreScan.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/StoreViewNodeStoreScan.java
#	community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/GraphDbStructureGuide.java
#	community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java
#	community/kernel/src/main/java/org/neo4j/storageengine/api/StoreReadLayer.java
#	community/kernel/src/main/java/org/neo4j/storageengine/api/schema/AbstractIndexReader.java
#	community/kernel/src/main/java/org/neo4j/storageengine/api/schema/IndexProgressor.java
#	community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableTransactionState.java
#	community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java
#	community/kernel/src/test/java/org/neo4j/graphdb/schema/DropBrokenUniquenessConstraintIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/api/CompositeIndexingIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/api/index/NodeUpdatesTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/api/schema/index/SchemaIndexDescriptorFactoryTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/BuiltInProceduresTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/LockingStatementOperationsTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulatorTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/ControlledPopulationIndexProvider.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexCRUDIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexMapTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexProxyAdapter.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexRecoveryIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexStatisticsIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexStatisticsTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexingServiceTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/inmemory/InMemoryIndexProvider.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/inmemory/UpdateCapturingIndexProvider.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingControllerTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/sampling/IndexSamplingJobTrackerTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/BuiltInProceduresIT.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/IndexQueryTransactionStateTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/IndexTxStateUpdaterTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/LabelTransactionStateTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/SchemaTransactionStateTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/api/store/SchemaCacheTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/DeferredConflictCheckingIndexUpdaterTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NativeUniqueSchemaIndexPopulatorTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/NumberLayoutTestUtil.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/StringLayoutTestUtil.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexProviderTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/newapi/IndexTxStateUpdaterTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/store/SchemaStorageTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/store/SchemaStoreTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/store/record/IndexRuleTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/store/record/SchemaRuleSerializationTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/store/record/SchemaRuleTestBase.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/Commands.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/NeoStoreTransactionApplierTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/NeoTransactionIndexApplierTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/PhysicalLogCommandReaderV2_2_4Test.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/state/DefaultIndexProviderMapTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/state/LogTruncationTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/state/SchemaRuleCommandTest.java
#	community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/state/TransactionRecordStateTest.java
#	community/kernel/src/test/java/org/neo4j/unsafe/batchinsert/internal/BatchInsertTest.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/LuceneIndexAccessor.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndex.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexProvider.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexProviderFactory.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/ReadOnlyDatabaseSchemaIndex.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/SchemaIndex.java
#	community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/WritableDatabaseSchemaIndex.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/index/LuceneSchemaIndexPopulationIT.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/index/LuceneSchemaIndexUniquenessVerificationIT.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/AccessUniqueDatabaseIndexTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexBuilderTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexCorruptionTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexPopulatorTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexProviderTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/LuceneSchemaIndexTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/populator/NonUniqueDatabaseIndexPopulatorTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/api/impl/schema/reader/PartitionedIndexReaderTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/impl/api/MultipleOpenCursorsTest.java
#	community/lucene-index/src/test/java/org/neo4j/kernel/impl/api/index/IndexingServiceIntegrationTest.java
#	community/neo4j/src/test/java/migration/Start3_2DbOn3_3AndCreateFusionIndexIT.java
#	community/neo4j/src/test/java/org/neo4j/locking/QueryExecutionLocksIT.java
#	community/neo4j/src/test/java/schema/MultiIndexPopulationConcurrentUpdatesIT.java
#	enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/EnterpriseCoreEditionModule.java
#	enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextTransactionEventUpdater.java
#	enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/FulltextUpdateApplier.java
#	enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/WritableFulltext.java
#	enterprise/fulltext-addon/src/test/java/org/neo4j/kernel/api/impl/fulltext/LuceneFulltextTestSupport.java
#	enterprise/fulltext-index/pom.xml
#	enterprise/ha/src/test/java/org/neo4j/kernel/api/SchemaIndexHaIT.java
#	enterprise/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/UniquenessConstraintCreationIT.java
#	integrationtests/src/test/java/org/neo4j/storeupgrade/StoreUpgradeIT.java
#	tools/src/main/java/org/neo4j/tools/dump/DumpCountsStore.java
#	tools/src/test/java/org/neo4j/tools/dump/DumpCountsStoreTest.java
#	tools/src/test/java/org/neo4j/tools/dump/InconsistencyReportReaderTest.java

ragadeeshu committed:
Merge branch '3.5-index-management-cleanup' into 3.4-fulltext-integration

ragadeeshu committed:
Fix some fulltext tests after merge
